---
title: "Untitled"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

```{r}
library(tfPaper)
library(mbtransfer)
library(tidyverse)
library(fs)
theme_set(my_theme())
```

```{r}
data_dir <- path("../../microbiome_transfer_data/david")
reads <- read_csv(data_dir / "abundance.csv") |>
  rename(sample = `...1`) |>
  column_to_rownames("sample") |>
  as.matrix()
samples <- read_csv(data_dir / "sample_metadata.csv", col_names = c("sample", "subject", "time")) |>
  mutate(intervention = str_extract(subject, "[A-z]+"))
subject <- read_csv(data_dir / "subject_data.csv")

keep_ix <- colMeans(reads != 0) > 0.7 # very aggressive, but just for testing
reads <- reads[, as.logical(keep_ix)]
```

```{r}
taxonomy <- read_csv("https://raw.githubusercontent.com/gerberlab/mitre/master/mitre/example_data/david/taxaTable.csv") |>
  mutate(
    name = str_c("Otu", str_pad(tax_id, 6, "left", "0")),
    sequence = NA,
  ) |>
  filter(name %in% colnames(reads)) |>
  rename(kingdom = superkingdom)
taxa_names <- colnames(reads)
new_ix <- !(taxa_names %in% taxonomy$name)
taxonomy <- taxonomy |>
  bind_rows(data.frame(name = taxa_names[new_ix]))
```


```{r}
interventions <- samples |>
  select(time, intervention, sample) |>
  mutate(value = 1) |>
  pivot_wider(names_from = intervention, values_fill = 0) |>
  column_to_rownames("sample") |>
  mutate(
    perturbation_window = time >= 0 & time <= 5,
    Animal = ifelse(perturbation_window & Animal == 1, 1, 0),
    Plant = ifelse(perturbation_window & Plant == 1, 1, 0)
  ) |>
  select(Plant, Animal)

subject <- subject |>
  rename(subject = subject_ID)
samples <- samples |>
  rename(condition = intervention)
```

```{r}
ts <- ts_from_dfs(reads, interventions, samples, subject) |>
  interpolate(method = "linear")

values_df <- map_dfr(ts, ~ bind_cols(rownames_to_column(data.frame(t(values(.))), "sample"), time = .@time), .id = "subject") |>
  pivot_longer(starts_with("Otu"), names_to = "taxon")
```

```{r, fig.height = 3, fig.width = 10}
top_taxa <- values_df |>
  group_by(taxon) |>
  summarise(mean = mean(value)) |>
  slice_max(mean, n = 7) |>
  pull(taxon)

values_df |>
  group_by(taxon) |>
  mutate(value = rank(value)) |>
  interaction_hm(subject, top_taxa, "diet")
```

### Cross Validation

Here is where we will compare against a few benchmarks.

### Identifying Perturbed Taxa

Here, we'll look at trajectories for a few taxa that are strongly/weakly
affected by the different intervention types. We'll also cluster the shapes of
different predicted trajectories, for a more nuanced view into short/long-term
effects.

```{r}
subject_data(ts) <- subject_data(ts) |>
  mutate(diet = ifelse(diet == "Plant", 0, 1))
fit <- mbtransfer(ts, P = 1, Q = 1)
```

```{r}
ts_missing <- subset_values(ts, 1:8)
ts_pred <- mbtransfer_predict(fit, ts_missing)
```

```{r}
ts_df <- pivot_ts(ts_pred) |>
  rename(y_hat = value) |>
  select(taxon, sample, time, y_hat) |>
  left_join(pivot_ts(ts)) |>
  rename(y = value) |>
  filter(y != y_hat)
```


```{r, fig.width = 8, fig.height = 8}
ts_df |>
  filter(taxon %in% unique(ts_df$taxon)[1:16]) |>
  ggplot() +
  geom_abline(slope = 1, col = "#F25A38") +
  geom_point(aes(y, y_hat), size = 0.5, alpha = 0.8) +
  scale_x_log10() +
  scale_y_log10() +
  facet_wrap(~ taxon, scales = "free")
```
