---
title: "Untitled"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

```{r}
library(tfPaper)
library(mbtransfer)
library(tidyverse)
library(fs)
library(tidymodels)
theme_set(my_theme())
```

```{r}
data_dir <- path("../../microbiome_transfer_data/david")
reads <- read_csv(data_dir / "abundance.csv") |>
  rename(sample = `...1`) |>
  column_to_rownames("sample") |>
  as.matrix()
samples <- read_csv(data_dir / "sample_metadata.csv", col_names = c("sample", "subject", "time")) |>
  mutate(intervention = str_extract(subject, "[A-z]+"))
subject <- read_csv(data_dir / "subject_data.csv")

keep_ix <- colMeans(reads != 0) > 0.8 # very aggressive, but just for testing
reads <- reads[, as.logical(keep_ix)]
```

```{r}
taxonomy <- read_csv("https://raw.githubusercontent.com/gerberlab/mitre/master/mitre/example_data/david/taxaTable.csv") |>
  mutate(
    name = str_c("Otu", str_pad(tax_id, 6, "left", "0")),
    sequence = NA,
  ) |>
  filter(name %in% colnames(reads)) |>
  rename(kingdom = superkingdom)
taxa_names <- colnames(reads)
new_ix <- !(taxa_names %in% taxonomy$name)
taxonomy <- taxonomy |>
  bind_rows(data.frame(name = taxa_names[new_ix]))
```


```{r}
interventions <- samples |>
  select(time, intervention, sample) |>
  mutate(value = 1) |>
  pivot_wider(names_from = intervention, values_fill = 0) |>
  column_to_rownames("sample") |>
  mutate(
    perturbation_window = time >= 0 & time <= 5,
    Animal = ifelse(perturbation_window & Animal == 1, 1, 0),
    Plant = ifelse(perturbation_window & Plant == 1, 1, 0)
  ) |>
  select(Plant, Animal)

subject <- subject |>
  rename(subject = subject_ID)
samples <- samples |>
  rename(condition = intervention)
```

```{r}
ts <- reads |>
  normalize("mbImpute", samples) |>
  ts_from_dfs(interventions, samples, subject) |>
  interpolate(method = "linear")
values_df <- pivot_ts(ts)
```

```{r, fig.height = 3, fig.width = 10, eval = FALSE}
top_taxa <- values_df |>
  group_by(taxon) |>
  summarise(mean = mean(value)) |>
  slice_max(mean, n = 7) |>
  pull(taxon)

values_df |>
  group_by(taxon) |>
  mutate(value = rank(value) / n()) |>
  interaction_hm(subject, top_taxa, "diet")
```

### Cross Validation

Here is where we will compare against a few benchmarks.

### Identifying Perturbed Taxa

Here, we'll look at trajectories for a few taxa that are strongly/weakly
affected by the different intervention types. We'll also cluster the shapes of
different predicted trajectories, for a more nuanced view into short/long-term
effects.

```{r}
subject_data(ts) <- subject_data(ts) |>
  mutate(diet = ifelse(diet == "Plant", 0, 1))
fit <- mbtransfer(ts, P = 2, Q = 2)
```

```{r, eval = FALSE}
ts_missing <- subset_values(ts, 1:8)
ts_pred <- predict(fit, ts_missing)
```

```{r, eval = FALSE}
ts_df <- pivot_ts(ts_pred) |>
  rename(y_hat = value) |>
  select(taxon, sample, time, y_hat) |>
  left_join(pivot_ts(ts)) |>
  rename(y = value) |>
  filter(y != y_hat)
```

```{r, fig.width = 8, fig.height = 8, eval = FALSE}
ts_df |>
  filter(taxon %in% unique(ts_df$taxon)[1:16]) |>
  ggplot() +
  geom_abline(slope = 1, col = "#F25A38") +
  geom_point(aes(y, y_hat), size = 0.5, alpha = 0.8) +
  scale_x_log10() +
  scale_y_log10() +
  facet_wrap(~ taxon, scales = "free")
```

### Inference of Significant Taxa

```{r}
ws <- steps(c("Plant" = TRUE, "Animal" = FALSE), lengths = 1:3, L = 4) |>
  c(steps(c("Plant" = FALSE, "Animal" = TRUE), lengths = 1:3, L = 4))

tr_fun <- function(x) {
  mbtransfer(x, P = 3, Q = 3, verbose = FALSE)
}

staxa <- select_taxa(ts, ws[[1]], ws[[8]], tr_fun, n_splits = 20)
#staxa <- readRDS("../data/staxa-diet.rds")
```

```{r}
ggplot(staxa$ms) +
  geom_point(aes(taxon, m)) +
  facet_wrap(~ lag, scales = "free_y")
```

```{r}
focus_taxa <- unlist(map(staxa$taxa, ~ c(.)))

focus_m <- staxa$ms |>
  mutate(taxon = taxa(ts)[taxon]) |>
  filter(taxon %in% focus_taxa) |>
  group_by(taxon, lag) |>
  summarise(m = mean(m)) |> 
  group_by(taxon) |>
  mutate(m = m / max(abs(m)))
  
focus_m_wide <- focus_m |>
  pivot_wider(names_from = lag, values_from = m) |>
  ungroup()

focus_m <- focus_m_wide |>
  select(-taxon) %>%
  recipe(~ ., data = .) |>
  step_pca() |>
  prep() |>
  juice() |>
  mutate(taxon = focus_m_wide$taxon) |>
  right_join(focus_m)

ggplot(focus_m) +
  geom_line(aes(lag, m, group = taxon, col = `2`))
```

* Get m's for the taxa within the selected set X
* Cluster the m's. X
* Fit a model to the full data and simulate differenced trajectories for the selected/clustered taxa X
* Visualize the differenced trajectories, faceted by cluster

```{r}
fit <- tr_fun(ts)
ws <- steps(c("Plant" = TRUE, "Animal" = FALSE), lengths = 1:3, L = 8) |>
  c(steps(c("Plant" = FALSE, "Animal" = TRUE), lengths = 1:3, L = 8))
sim_ts <- counterfactual_ts(ts, ws[[1]], ws[[8]], start_ix = 4) |>
  map(~ predict(fit, .))
```

```{r}
ribbon_data <- (sim_ts[[2]] - sim_ts[[1]]) |>
  pivot_ts() |>
  filter(taxon %in% focus_taxa) |>
  group_by(taxon, time) |>
  summarise(
    q25 = quantile(value, 0.25),
    median = median(value),
    q75 = quantile(value, 0.75)
  ) |>
  left_join(focus_m)
```

```{r, fig.height = 7, fig.width = 10}
ggplot(ribbon_data, aes(x = time)) +
  geom_ribbon(aes(ymin = q25, ymax = q75), fill = "#d3d3d3") +
  geom_hline(yintercept = 0, size = .25, col = "#787878") +
  geom_line(aes(y = median), col = "#8e8e8e", size = 1) +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 8)) +
  facet_wrap(~ reorder(taxon, `2`), scales = "free_y") +
  theme(axis.text = element_text(size = 8))
```

```{r, fig.height = 5, fig.width = 10}
values_df |>
  filter(diet == "Animal") |>
  group_by(taxon) |>
  mutate(value = rank(value) / n()) |>
  interaction_hm(taxa = c("Otu000095", "Otu000039", "Otu000051"), "diet")
```
