---
title: "Untitled"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r}
library(microTF)
library(tidyverse)
library(fs)
theme_set(theme_bw())
reticulate::use_condaenv("~/miniforge-pypy3/envs/mdsine2/")
```


```{r}
ts_to_df <- function(ts, tax_ix=17) {
  map_dfr(
    ts, 
    ~ tibble(
      time = .@time, 
      value = values(.)[tax_ix, ],
      plant = interventions(.)["Plant", ],
      animal = interventions(.)["Animal", ]
    ),
    .id = "subject"
  ) %>%
    mutate(subject = str_c("S", subject))
}

intervention_plot <- function(ts_df) {
  ggplot(ts_df, aes(time, reorder(subject, value))) + 
    geom_tile(aes(fill = value)) +
    geom_point(data = ts_df %>% filter(animal == 1), shape = 1) +
    scale_fill_distiller(direction = 1) +
    labs(title = tax_ix)
}
```

### Data Preprocessing

```{r}
data_dir <- path("../../microbiome_transfer_data/david")
abundances <- read_csv(data_dir / "abundance.csv") %>%
  rename(sample = `...1`)
samples <- read_csv(data_dir / "sample_metadata.csv", col_names = c("sample", "subject", "time")) %>%
  mutate(intervention = str_extract(subject, "[A-z]+"))
subject <- read_csv(data_dir / "subject_data.csv")
keep_ix <- colMeans(as.matrix(abundances[, -1]) != 0) > 0.2
reads <- abundances[, c(TRUE, keep_ix)] %>%
  column_to_rownames("sample") %>%
  .[, 1:25] %>%
  asinh()
```


```{r}
interventions <- samples %>%
  select(time, intervention, sample) %>%
  mutate(value = 1) %>%
  pivot_wider(names_from = intervention, values_fill = 0) %>%
  column_to_rownames("sample") %>%
  mutate(
    perturbation_window = time >= 0 & time <= 5,
    Animal = ifelse(perturbation_window & Animal == 1, 1, 0),
    Plant = ifelse(perturbation_window & Plant == 1, 1, 0) 
  ) %>%
  select(Plant, Animal)

ts <- ts_from_dfs(reads, interventions, samples) %>%
  interpolate()
```

### Original trajectories

We first visualize some of the original trajectories.

```{r}
p <- list()
for (tax_ix in seq_len(25)) {
  p[[tax_ix]] <- intervention_plot(ts_to_df(ts, tax_ix))
} 
```

```{r}
p[[5]]
p[[11]]
p[[17]]
```

### Visualizing Counterfactuals

Next, let's visualize the counterfacutal where we assume everyone was / was not
given the intervention. For this proof of concept, we're not going to bother
splitting the data into train / test groups.

```{r}
fit <- train(ts, "gbm")
```

We will mask all the timepoints before the first intervention, and we will
manipulate the interventions manually.

```{r}
ts0 <- ts
ts1 <- ts

for (i in seq_along(ts)) {
  inter <- interventions(ts[[i]])
  init_ix <- min(which(colSums(inter) > 0))
  v <- values(ts[[i]])[, seq_len(init_ix - 1)]
  
  values(ts0[[i]]) <- v
  values(ts1[[i]]) <- v
  
  inter[, seq(init_ix, ncol(inter))] <- 0
  interventions(ts0[[i]]) <- inter
  
  inter["Animal", seq(init_ix, init_ix + 3)] <- 1
  interventions(ts1[[i]]) <- inter
}

ts0 <- predict(fit, ts0)
ts1 <- predict(fit, ts1)

```


```{r}
tax_ix <- 17
ts1_df <- ts_to_df(ts1, tax_ix)
ggplot(ts1_df) +
  geom_histogram(aes(value, fill = as.factor(animal)), position = "identity", alpha = 0.8) +
  scale_fill_brewer(palette = "Set1")

intervention_plot(ts1_df)
```

Here is the same plot under the counterfactual that no one was intervened upon.

```{r}
ts0_df <- ts_to_df(ts0, tax_ix)
ggplot(ts0_df) +
  geom_histogram(aes(value, fill = as.factor(animal)), position = "identity", alpha = 0.8) +
  scale_fill_brewer(palette = "Set1")
intervention_plot(ts0_df)
```

We can plot the differences too.

```{r}
ts_counter <- ts1_df %>%
  rename(value1 = value) %>%
  left_join(
    ts0_df %>%
      rename(value0 = value) %>%
      select(-animal, -plant)
  )
```

There is some evidence of a delayed effect.

```{r}
ggplot(ts_counter) +
  geom_tile(aes(time, reorder(subject, abs(value1 - value0)), fill = value1 - value0)) +
  scale_fill_gradient2()
```

Let's make the analogous figure for a taxon that we think have negative /
neutral effects.

```{r}
ts_counter <- ts_to_df(ts1, 11) %>%
  rename(value1 = value) %>%
  left_join(
    ts_to_df(ts0, 11) %>%
      rename(value0 = value) %>%
      select(-animal, -plant)
  )

ggplot(ts_counter) +
  geom_tile(aes(time, reorder(subject, abs(value1 - value0)), fill = value1 - value0)) +
  scale_fill_gradient2()

ts_counter <- ts_to_df(ts1, 5) %>%
  rename(value1 = value) %>%
  left_join(
    ts_to_df(ts0, 5) %>%
      rename(value0 = value) %>%
      select(-animal, -plant)
  )

ggplot(ts_counter) +
  geom_tile(aes(time, reorder(subject, abs(value1 - value0)), fill = value1 - value0)) +
  scale_fill_gradient2()

```

Note that training performance is sensitive to the transformation we gave. We
are not looking into the full confidence band of each series, but this could be
of potential interest.

This figure is optimistic, because we haven't split the training and impulse
function visualization. As a summary, it is okay to look at the full data. But
as soon as we want to have some population level guarantee on the difference
between counterfactual transfer functions, this will not be sufficient.

I would like to show this on the same time scale (i.e., relative to min
intervention time, so they are more aligned). It would be useful to automate the
calculation of the difference (put it in a package function) and show it as a
simplified line plot across a collection of taxa.

Finally, we want to be able to learn variations of the function across
populations.