---
title: "Untitled"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r}
library(tidyverse)
library(reticulate)
library(fs)
theme_set(theme_bw())
use_condaenv("~/miniconda3/envs/mdsine2")
```

There might actually be some value in supporting a testing procedure. The
intervention effects are clear for some taxa, but certainly do not seem
universal.

```{r}
data_dir <- path("../../microbiome_transfer_data/david")
abundances <- read_csv(data_dir / "abundance.csv") %>%
  rename(sample = `...1`)
samples <- read_csv(data_dir / "sample_metadata.csv", col_names = c("sample", "group", "time")) %>%
  mutate(intervention = str_extract(group, "[A-z]+"))
subject <- read_csv(data_dir / "subject_data.csv")
taxonomy <- read_csv(data_dir / "taxaTable.csv")

keep_ix <- colMeans(as.matrix(abundances[, -1]) != 0) > 0.2
abundances <- abundances[, c(TRUE, keep_ix)]
```

```{r, fig.width = 20, fig.height = 7.5}
x_sub <- abundances[, 1:26] %>%
  pivot_longer(-sample, names_to = "taxon") %>%
  group_by(sample) %>%
  mutate(value = value / sum(value)) %>%
  group_by(taxon) %>%
  mutate(group_max = max(value)) %>%
  left_join(samples)

ggplot(x_sub, aes(time, value, col = intervention)) +
  geom_rect(aes(xmin = 0, ymin = 0, xmax = 5, ymax = group_max), fill = "#D9D9D9", col = NA, alpha = 0.2) +
  geom_line(aes(group = group), alpha = 0.8) +
  stat_smooth(se = FALSE, size = 2) +
  scale_color_manual(values = c("#D99066", "#358C6C")) +
  scale_y_sqrt() +
  facet_wrap(~ reorder(taxon, value), scales = "free_y", ncol = 5)
```

```{r}
# mdsine crashes if 0 is not one of the measured timepoints
subjects <- unique(samples$group)
for (i in seq_along(subjects)) {
  current_times <- samples %>%
    filter(group == subjects[i]) %>%
    pull(time)
  
  if (!(0 %in% current_times)) {
    replace_ix <- max(which(samples$group == subjects[i] & samples$time <= 0))
    samples$time[replace_ix] <- 0
  }
}

metadata <- samples %>%
  rename(sampleID = sample, subject = group) %>%
  select(-intervention)

qpcr <- metadata %>%
  mutate(
    measurement1 = rnorm(nrow(metadata), 1e9, 1e3), 
    measurement2 = rnorm(nrow(metadata), 1e9, 1e3), 
    measurement3 = rnorm(nrow(metadata), 1e9, 1e3)
  ) %>%
  select(sampleID, measurement1:measurement3)

perturbations <- samples %>%
  group_by(group, intervention) %>%
  summarise(
    start = max(time[time <= 0]),
    end = min(time[time >= 5])
  ) %>%
  rename(subject = group, name = intervention)

subject_ids <- samples %>%
  pull(group) %>%
  unique()
perturbation_ids <- samples %>%
  pull(intervention) %>%
  unique()
perturbation_pad <- expand.grid(subject_ids, perturbation_ids) %>%
  mutate(
    start = 0,
    end = 1 
  ) %>%
  rename(name = Var2, subject = Var1) %>%
  filter(!str_detect(subject, as.character(name)))

perturbations <- bind_rows(perturbations, perturbation_pad)
```


```{r}
reads <- abundances %>%
  filter(sample %in% metadata$sampleID) %>%
  pivot_longer(-sample, names_to = "name") %>%
  pivot_wider(names_from = "sample")

taxonomy <- taxonomy %>%
  mutate(
    name = str_c("Otu", str_pad(tax_id, 6, "left", "0")),
    sequence = NA,
  ) %>%
  filter(name %in% reads$name) %>%
  rename(kingdom = superkingdom)

reads <- reads %>%
  filter(name %in% taxonomy$name)

write_tsv(metadata, "metadata.tsv")
write_tsv(qpcr, "qpcr.tsv")
write_tsv(perturbations, "perturbations.tsv")
write_tsv(reads, "reads.tsv")
write_tsv(taxonomy, "taxonomy.tsv")
```

